### ISSUE TITLE
Trace.copy: Using deepcopy causes derived Stats keys setting parent keys with computed values (leading to wrong values in edge cases)

### ISSUE THREAD

**Issue by cjhopp** [obspy/obspy#2600]
Hi everybody,

First, I hope all of you are happy and healthy.

Second, I've got a puzzling issue when working with highly-sampled data:

-  ObsPy version: 1.2.1 (anaconda)
-  Python version: 3.7
-  Arch Linux

Working with sampling rates >= 100000. Hz produces unexpected behavior.

```
tr = Trace(data=np.ones(100000), header={'sampling_rate': 100000.})
print(tr.stats.sampling_rate)
100000.0
print(tr.copy().stats.sampling_rate)
99999.99999999999
```
I think this is a result of floating point ["representation error"](https://docs.python.org/3/tutorial/floatingpoint.html), whenever the Trace AttribDict needs to be re-evaluated, [specifically where `delta` is involved](https://github.com/obspy/obspy/blob/88687a146a7c3ca8f35c608db2243cf5fed6813c/obspy/core/trace.py#L177-L182).

Not sure how to go about addressing this though...

**Comment by cjhopp** [comment]
Changing [this](https://github.com/obspy/obspy/blob/88687a146a7c3ca8f35c608db2243cf5fed6813c/obspy/core/trace.py#L180) line so that it forces a rounded float:

`value = float(round(1.0 / float(value)))`

fixes my issue but would obviously break anything with a non-integer sampling rate.

Are there use-cases where decimal sampling rates are required? If so, I'm open to suggestions as to how to accommodate >100000 Hz sampling rates without rounding. Otherwise, I'm happy to submit a PR with the above change (including a little test if that's necessary?).

**Comment by d-chambers** [comment]
Hi @cjhopp,

Interesting issue! Unfortunately your problem involves parts of ObsPy that, if changed, could have huge implications for most users' codes so we need to be very very careful.

I don't think rounding to an integer value is a good idea. I have personally worked with some obscure digitizers that sample at non integer intervals and we don't want to (retroactively) preclude their use. 

The only potential path forward (which I can think of) is to leave current support for floats alone and somehow support integer representations as well. Off the top of my head, maybe stats could avoid changing `sampling_rate` datatypes at all? You could then get int or float values depending on what was provided.

Another approach might be to support `numpy.timedelta64` in some fashion for `time_delta` so a very precise sampling_period can be specified, even if the complementary `sampling_rate` is not a whole number when expressed in Hz.

**Comment by cjhopp** [comment]
Hi @d-chambers,

Thanks for the guidance! I figured there must have been a need for float values. I've just never encountered such a thing.

I like the idea of returning the type that the user provides. This might still have implications for codes where `sampling_rate` is provided as an integer on creation of the `Trace`, but then assumed to be a float for later operations?

The use case that brought me to this issue is a sanity check in the depths of `eqcorrscan`:
```
if tr.stats.sampling_rate * length != tr.stats.npts:
```
where `length` is some number of seconds. Here, the assumption is that the integer `npts` is a multiple of `sampling_rate`, which isn't so for the example above. Perhaps on the user end it's better not to assume `sampling_rate` is a whole number, but I expect these sorts of statements that assume it _is_ are fairly ubiquitous in users codebases.

So I think integer support for `sampling_rate` is a good idea, maybe also with a warning in cases where the user provides `sampling_rate >= 100000.` as a float, stating that it may produce unexpected results?

**Comment by d-chambers** [comment]
Would it be easier to just loosen some of the eqcorrscan checks? Maybe something like this would work?

```python
if not np.isclose(np.round(tr.stats.sampling_rate * length), tr.stats.npts)
``` 

If you want to make a fix on the ObsPy level I would personally start by adding a few tests:

1. A test to make sure the sampling rate doesn't change when copying a trace (like in your example)

2. A test for sampling rates which aren't whole numbers

3. A test that int sampling rates stay ints after assignment 

I haven't looked at the `Stats` internals in awhile so I don't know how hard it will be, but I have seen enough of your code in eqcorrscan to know you are a very capable developer. Also, I am happy to help if you need it! :+1: 

I wouldn't recommend any new warnings though, as the amount of numerical precision accumulated (at least in your example) is going to be array length dependent, so there is not a magic number we can use to decide exactly when to raise a warning or not.

**Comment by cjhopp** [comment]
It's certainly something I'll take up with @calum-chamberlain over at eqcorrscan.

But I do think it's something worth addressing at the obspy level, although I acknowledge that it's a very edge case at this point.

I'll take your test suggestions and see where I get in the next week or so!

**Comment by megies** [comment]
@d-chambers @cjhopp so I briefly looked into this.. the problem seems to be that we are lazy and use `deepcopy` on the whole `Trace` and then when it copies `Stats` it actually sets all attributes in it, even some of the derived ones like `delta`. So the reason for that is that we did not define `delta` as readonly, even though it itself is not stored but only `sampling_rate`, basically setting `sampling_rate` with the inverse of the value provided for setting `delta`.

Now, while this should even still be OK in most cases, @cjhopp you run into a very strange Python floating point accuracy edge case:

```python
In [2]: 1 / 1e-05                                                                                                                         
Out[2]: 99999.99999999999

In [3]: 1 / 1e-06                                                                                                                         
Out[3]: 1000000.0

In [4]: 1 / 1e-07                                                                                                                         
Out[4]: 10000000.0

In [5]: 1 / 1e-08                                                                                                                         
Out[5]: 100000000.0

In [6]: 1 / 1e-04                                                                                                                         
Out[6]: 10000.0

In [7]: 1 / 1e-05                                                                                                                         
Out[7]: 99999.99999999999

In [8]: 1/1e-5                                                                                                                            
Out[8]: 99999.99999999999

In [9]: 1e5 == 1 / 1e-5                                                                                                                   
Out[9]: False
```

---

### Long story short

In principle it's not our problem, but we could get around it by being less lazy in `Trace.copy` and only use `deepcopy` for some parts and explicitly skip setting derived keys in `Stats` (here: `delta`), thus avoiding to set the parent keys (here: `sampling_rate`) with derived values computed from derived ones (here: `1 / float(1e-5)` evaluating to 99999.9).

**Comment by trichter** [comment]
In the past, I wondered if the simple deepcopy of the stats with its derived values just works for all cases. This issue shows, that the answer is No. I will have a look into it.

**Comment by cjhopp** [comment]
@megies I didn't realize how edge-y my case really was! Maybe we should just budget for more disks and record at 1 MHz instead...

@trichter Thank you for having a look. Let me know if there is anything I can do to help.

**Comment by trichter** [comment]
@cjhopp You could check if the changes in the linked PR solve your issues.
(Install obspy from this branch or simply `pip install https://github.com/obspy/obspy/archive/fix_deepcopy.zip`)

**Comment by cjhopp** [comment]
Yep, no more problems on my end. Thanks again for the quick fix @trichter !
